File: flutter_app/lib/utils/reference_utils.dart
import '../models/reference.dart';

enum SmartActionType { phone, email, url, address }

class SmartAction {
  final SmartActionType type;
  final String value;
  final String label;
  final String actionUrl;

  const SmartAction({
    required this.type,
    required this.value,
    required this.label,
    required this.actionUrl,
  });
}

class ReferenceUtils {
  /// Extract phone numbers, URLs, emails, and physical addresses from content
  static List<SmartAction> detectSmartActions(String content) {
    if (content.trim().isEmpty) return [];

    final List<SmartAction> actions = [];
    final trimmed = content.trim();

    // 1. Phone numbers
    final phoneRegex = RegExp(r'(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}|\+?\d{1,3}[-.\s]?\d{4,5}[-.\s]?\d{4,6}');
    final phoneMatches = phoneRegex.allMatches(trimmed);
    for (final match in phoneMatches) {
      final phone = match.group(0)?.trim() ?? '';
      final digits = phone.replaceAll(RegExp(r'[^\d+]'), '');
      if (digits.length >= 7 && digits.length <= 15) {
        if (!actions.any((a) => a.value == phone)) {
          actions.add(SmartAction(
            type: SmartActionType.phone,
            value: phone,
            label: 'Call $phone',
            actionUrl: 'tel:${phone.replaceAll(RegExp(r'\s+'), '')}',
          ));
        }
      }
    }

    // 2. URLs
    final urlRegex = RegExp(r'(?:https?:\/\/|www\.)[^\s/$.?#].[^\s]*', caseSensitive: false);
    final urlMatches = urlRegex.allMatches(trimmed);
    for (final match in urlMatches) {
      final url = match.group(0)?.trim() ?? '';
      final href = url.startsWith('http://') || url.startsWith('https://') ? url : 'https://$url';
      if (!actions.any((a) => a.value == url)) {
        final labelUrl = url.length > 25 ? '${url.substring(0, 22)}...' : url;
        actions.add(SmartAction(
          type: SmartActionType.url,
          value: url,
          label: 'Open $labelUrl',
          actionUrl: href,
        ));
      }
    }

    // 3. Email addresses
    final emailRegex = RegExp(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}');
    final emailMatches = emailRegex.allMatches(trimmed);
    for (final match in emailMatches) {
      final email = match.group(0)?.trim() ?? '';
      if (!actions.any((a) => a.value == email)) {
        actions.add(SmartAction(
          type: SmartActionType.email,
          value: email,
          label: 'Email $email',
          actionUrl: 'mailto:$email',
        ));
      }
    }

    // 4. Physical Address Detection (conservative)
    final addressKeywords = RegExp(r'\b(Street|St\.?|Road|Rd\.?|Avenue|Ave\.?|Boulevard|Blvd\.?|Lane|Ln\.?|Drive|Dr\.?|Court|Ct\.?|Highway|Hwy\.?|Nagar|Marg|Sector|Plot|Floor|Apt|Suite|Apartment|Building)\b', caseSensitive: false);
    final lines = trimmed.split('\n');
    for (final line in lines) {
      final trimmedLine = line.trim();
      if (trimmedLine.length >= 10 && addressKeywords.hasMatch(trimmedLine)) {
        if (!trimmedLine.startsWith('http') && !trimmedLine.contains('@') && !actions.any((a) => a.type == SmartActionType.address)) {
          final query = Uri.encodeComponent(trimmedLine);
          final labelAddr = trimmedLine.length > 20 ? '${trimmedLine.substring(0, 18)}...' : trimmedLine;
          actions.add(SmartAction(
            type: SmartActionType.address,
            value: trimmedLine,
            label: 'Map "$labelAddr"',
            actionUrl: 'https://www.google.com/maps/search/?api=1&query=$query',
          ));
          break;
        }
      }
    }

    return actions;
  }

  /// Format full reference for clipboard copy
  static String formatReferenceForCopy(Reference reference) {
    final parts = <String>[];
    if (reference.title.trim().isNotEmpty) {
      parts.add(reference.title.trim());
    }
    if (reference.content.trim().isNotEmpty) {
      parts.add(reference.content.trim());
    }
    if (reference.tags.isNotEmpty) {
      parts.add(reference.tags.join(' '));
    }
    return parts.join('\n');
  }

  /// Extract tags (+tag or @tag) from text
  static List<String> extractTagsFromText(String text) {
    if (text.isEmpty) return [];
    final words = text.split(RegExp(r'\s+'));
    final tags = <String>{};

    for (final word in words) {
      if ((word.startsWith('+') || word.startsWith('@')) && word.length > 1) {
        tags.add(word);
      }
    }

    return tags.toList();
  }
}

File: utils/referenceUtils.ts
import { Reference, ReferenceSmartAction } from '@/types/todo';

/**
 * Extract phone numbers, URLs, emails, and addresses from text cleanly and conservatively
 */
export function detectSmartActions(content: string): ReferenceSmartAction[] {
  if (!content || typeof content !== 'string') return [];

  const actions: ReferenceSmartAction[] = [];
  const trimmed = content.trim();

  // 1. Phone numbers
  // Matches international format (+91 98765 43210, +1-555-123-4567) or standard formats (9876543210, 555-123-4567, (555) 123-4567)
  const phoneRegex = /(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}|\+?\d{1,3}[-.\s]?\d{4,5}[-.\s]?\d{4,6}/g;
  const phoneMatches = trimmed.match(phoneRegex);
  if (phoneMatches) {
    for (const match of phoneMatches) {
      const cleanPhone = match.trim();
      const digitsOnly = cleanPhone.replace(/[^\d+]/g, '');
      if (digitsOnly.length >= 7 && digitsOnly.length <= 15) {
        if (!actions.some(a => a.value === cleanPhone)) {
          actions.push({
            type: 'phone',
            value: cleanPhone,
            label: `Call ${cleanPhone}`,
            actionUrl: `tel:${cleanPhone.replace(/\s+/g, '')}`
          });
        }
      }
    }
  }

  // 2. URLs
  const urlRegex = /(?:https?:\/\/|www\.)[^\s/$.?#].[^\s]*/gi;
  const urlMatches = trimmed.match(urlRegex);
  if (urlMatches) {
    for (const match of urlMatches) {
      const cleanUrl = match.trim();
      const href = cleanUrl.startsWith('http://') || cleanUrl.startsWith('https://')
        ? cleanUrl
        : `https://${cleanUrl}`;
      if (!actions.some(a => a.value === cleanUrl)) {
        actions.push({
          type: 'url',
          value: cleanUrl,
          label: `Open ${cleanUrl.length > 25 ? cleanUrl.substring(0, 22) + '...' : cleanUrl}`,
          actionUrl: href
        });
      }
    }
  }

  // 3. Email addresses
  const emailRegex = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;
  const emailMatches = trimmed.match(emailRegex);
  if (emailMatches) {
    for (const match of emailMatches) {
      const cleanEmail = match.trim();
      if (!actions.some(a => a.value === cleanEmail)) {
        actions.push({
          type: 'email',
          value: cleanEmail,
          label: `Email ${cleanEmail}`,
          actionUrl: `mailto:${cleanEmail}`
        });
      }
    }
  }

  // 4. Physical Address Detection (conservative)
  // Look for common address keywords (Road, Street, Ave, St, Rd, Blvd, Lane, Marg, Nagar, Bandra, etc.)
  const addressKeywords = /\b(Street|St\.?|Road|Rd\.?|Avenue|Ave\.?|Boulevard|Blvd\.?|Lane|Ln\.?|Drive|Dr\.?|Court|Ct\.?|Highway|Hwy\.?|Nagar|Marg|Sector|Plot|Floor|Apt|Suite|Apartment|Building)\b/i;
  const lines = trimmed.split('\n');
  for (const line of lines) {
    const trimmedLine = line.trim();
    if (trimmedLine.length >= 10 && addressKeywords.test(trimmedLine)) {
      // Check if it's not a pure URL or email
      if (!trimmedLine.startsWith('http') && !trimmedLine.includes('@') && !actions.some(a => a.type === 'address')) {
        const safeQuery = encodeURIComponent(trimmedLine);
        actions.push({
          type: 'address',
          value: trimmedLine,
          label: `Map "${trimmedLine.length > 20 ? trimmedLine.substring(0, 18) + '...' : trimmedLine}"`,
          actionUrl: `https://www.google.com/maps/search/?api=1&query=${safeQuery}`
        });
        break;
      }
    }
  }

  return actions;
}

/**
 * Format full reference for clipboard copy
 */
export function formatReferenceForCopy(reference: Reference): string {
  const parts: string[] = [];
  if (reference.title && reference.title.trim()) {
    parts.push(reference.title.trim());
  }
  if (reference.content && reference.content.trim()) {
    parts.push(reference.content.trim());
  }
  if (reference.tags && reference.tags.length > 0) {
    parts.push(reference.tags.join(' '));
  }
  return parts.join('\n');
}

/**
 * Extract tags (+project or @context or general tags) from text
 */
export function extractTagsFromText(text: string): string[] {
  if (!text) return [];
  const words = text.split(/\s+/);
  const tags = new Set<string>();

  for (const word of words) {
    if ((word.startsWith('+') || word.startsWith('@')) && word.length > 1) {
      tags.add(word);
    }
  }

  return Array.from(tags);
}
